index.bs

```diff
@@ -1498,14 +1526,43 @@ an assertion. Its {{PublicKeyCredentialRequestOptions/challenge}} member must be
     ::  This optional member contains a list of {{PublicKeyCredentialDescriptor}} object representing [=public key credentials=]
         acceptable to the caller, in decending order of the caller's preference (the first item in the list is the most
         preferred credential, and so on down the list).
-    
+
+    :   <dfn>requireUserVerification</dfn>
+    ::  This member describes the [=[RP]=]'s requirements regarding [=user verification=] for the
+        {{CredentialsContainer/get()}} operation. Eligible authenticators are filtered to only those capable of satisfying this
+        requirement.
+
     :   <dfn>extensions</dfn>
     ::  This optional member contains additional parameters requesting additional processing by the client and authenticator.
         For example, if transaction confirmation is sought from the user, then the prompt string might be included as an
         extension.
 </dl>
 
 
+### User Verification Requirement enumeration (enum <dfn enum>UserVerificationRequirement</dfn>) ### {#userVerificationRequirement}
+
+<pre class="idl">
+    enum UserVerificationRequirement {
+        "required",
+        "preferred",
+        "discouraged"
+    };
+</pre>
+
+A [=[RP]=] may require [=user verification=] for some of its operations but not for others, and may use this type to express this
+level of requirement.
+
+The value {{UserVerificationRequirement/required}} indicates that the [=[RP]=] requires [=user verification=] for this operation
+and will fail the operation if the response does not have the [=UV=] [=flag=] set.
+
+The value {{UserVerificationRequirement/preferred}} indicates that the [=[RP]=] would like [=user verification=] for this
+operation if possible, but will not fail the operation if the response does not have the [=UV=] [=flag=] set.
+
+The value {{UserVerificationRequirement/discouraged}} indicates that the [=[RP]=] does not require [=user verification=], and does
+not want the [=client=] or [=authenticator=] to request it, but will not fail the operation if the response has the [=UV=]
+[=flag=] set.
+
```

Comment:
Why is it important to signal to the Authenticator whether or not the RP will fail the operation?

It seems to me that we can get the same semantics with a boolean field:

- An RP who intents to fail the operation if the UV bit is not set in the response will set the userVerificationRequirement to `true`, and then fail the operation if the UV bit is missing in the response.

- An RP who prefers user verification, but won't fail the operation if user verification is not performed, will set the userVerificationRequirement to `true`, and then not fail the operation if the UV bit is missing in the response.

- An RP who prefers to *not* do user verification, but won't fail the operation if user verification performed, will set the userVerificationRequirement to `false`, and then not fail the operation if the UV bit is present in the response.

So I would recommend to make this a boolean. It covers all the use cases described here, uses fewer values, and makes it much easier to map these values to the boolean option we have for this in CTAP. 

I would also recommend to name this `userVerification`.

Reply:
I agree with renaming to `userVerification`.

It's not important to signal to the authenticator, but it's useful to signal this to the client so it can select an appropriate authenticator if multiple ones are available.

- If `userVerification` is `"required"`, the client will send the command only to UV capable authenticators
- If `userVerification` is `"preferred"`, the client will send the command only to UV capable authenticators if any are available, and to any eligible authenticator otherwise
- If `userVerification` is `"discouraged"`, the client could refrain from sending the command to authenticators known to always require UV (i.e. first-factor-only authenticators).

If this were a Boolean parameter and `true` could mean either `"required"` or `"preferred"`, the client could fail the operation unnecessarily because it thinks no eligible authenticator is available.

Reply:
What's the benefit of doing it this way? Contrast this with the alternate approach:

- if `userVerification` is set to `true`, client will prioritize UV-capable Authenticators (send them the requests first, or hold responses from other Authenticators until it hears from an UV-capable Authenticator, etc.), but will accept responses from non-UV-capable Authenticators.

- if 'userVerification` is set to `false` (or missing), then the client will not prioritize according to UV capabilities.

As far as I can tell, the difference between this approach and the one you're describing is *when* the error occurs if

- the RP requires user verification, *and*
- there are non-UV capable Authenticators attached, *and*
- those non-UV-capable Authenticators manage credentials that are on the `allowCredentials` list (or there is no `allowCredentials` list), and would therefore process the request.

If all these conditions are met, then with the tri-state value, the client can fail the operation early, whereas in the boolean case, the RP will have to fail the operation. This will likely result in the user seeing an error a few hundred milliseconds later than they would in the tri-state case. 

Is that worth the added complexity?

Reply:
It becomes more relevant now that we are moving to waiting for hot-plugged authenticators; it's more likely a question of tens of seconds rather than hundreds of milliseconds.

In the hot-plugging context it's likely that no authenticators will be available at the start, and sometime (delay on the order of seconds) within the timeout window an authenticator will appear. If the RP has set `requireUserVerification: true` and this first authenticator doesn't support UV, should the client wait X more seconds to see if a UV capable authenticator appears? How should you design the UX around this so users aren't confused and start over because "it hangs"?

With the tri-state approach there is no ambiguity: If the RP sets `userVerification: "required"` and the first authenticator doesn't support UV, the client can immediately tell the user to use another authenticator. If the RP sets `userVerification: "preferred"` the client knows it doesn't have to wait for a UV capable authenticator and can immediately go with whatever the user provides.

I also find it cognitively jarring that "required" wouldn't actually mean *required*, but I guess that's more of a subjective aesthetic issue.

All that said, you're right that it might not be worth the complexity if we assume that most users will only have one authenticator and likely won't ever find themselves in this situation. On the other hand this would save RPs the complexity of informing the user of this failure mode, since that would be handled by the browser instead. I think it's reasonable to make implementing an RP easier by pushing this to the browser, but then again I'm not the one who would be implementing it. :)

@AngeloKai @jcjones thoughts on this?

Reply:
Also, addressing this:

>[a Boolean parameter instead] makes it much easier to map these values to the boolean option we have for this in CTAP.

The mapping is already suggested in the spec of `create()` and `getAssertion()`; does it need to be made clearer?